File: src/test/java/org/ymxs/CsvTest.java
package org.ymxs;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;
import org.ymxs.YMXS.Multi;
import org.ymxs.YMXS.Prescaler;
import org.ymxs.YMXS.Register;
import org.ymxs.YMXS.Row;
import org.ymxs.YMXS.Source;
import org.ymxs.YMXS.Start;
import org.ymxs.YMXS.Timer;
import org.ymxs.YMXS.Tune;

/**
 * CSV against JSON. Both write the same structure, so a tune written in
 * one form and read from the other is an equal tune.
 */
final class CsvTest {

    private static List<Path> tunes() throws IOException {
        try (Stream<Path> at = Files.list(Path.of("doc/tunes"))) {
            return at.filter(one -> one.toString().endsWith(".json")).sorted().toList();
        }
    }

    @Test
    void everyTuneCrossesFromOneFormToTheOther() throws IOException {
        List<Path> tunes = tunes();
        assertTrue(tunes.size() >= 4, () -> "only " + tunes.size() + " tunes read");
        for (Path at : tunes) {
            Multi text = Text.read(Files.readString(at));
            assertEquals(text, Csv.read(Csv.write(text)),
                    at + " is not the multi it was, written as a table and read back");
        }
    }

    @Test
    void theCsvFormReadsBackAsItStands() throws IOException {
        Path at = Path.of("doc/tunes/circus.csv");
        String csv = Files.readString(at);
        assertEquals(csv, Csv.write(Csv.read(csv)), at + " does not write back as it reads");
        assertEquals(Text.read(Files.readString(Path.of("doc/tunes/circus.json"))),
                Csv.read(csv), "the two forms of one tune read into one structure");
    }

    @Test
    void aHashLineNamesATableAndTheLineAfterItNamesTheColumns() throws IOException {
        List<String> lines = Files.readString(Path.of("doc/tunes/circus.csv")).lines()
                .toList();
        List<String> named = lines.stream().filter(one -> one.startsWith("###"))
                .map(one -> one.substring(3).strip()).toList();
        assertEquals(List.of("multi", "tune", "rows"), named,
                "circus is rows alone, so those are the three tables it opens");
        int rows = lines.indexOf("### rows");
        assertEquals(List.of("row", "r0", "r1", "r2", "r3", "r4", "r5", "r6",
                "r7", "r8", "r9", "r10", "r11", "r12", "r13"),
                Csv.cells(lines.get(rows + 1)),
                "a row of a tune is a row here, with a column a register");
    }

    /** The point of the two lines: a column name stands over the cells it
     *  names, so a reader counts the columns of a table by reading down. */
    @Test
    void aColumnNameStandsOverTheCellsItNames() throws IOException {
        List<String> lines = Files.readString(Path.of("doc/tunes/circus.csv")).lines()
                .toList();
        for (int at = 0; at < lines.size(); at++) {
            if (!lines.get(at).startsWith("###")) {
                continue;
            }
            String named = lines.get(at).substring(3).strip();
            assertTrue(!named.contains(","), "a heading line names the table alone: " + named);
            int columns = Csv.cells(lines.get(at + 1)).size();
            for (int row = at + 2; row < lines.size() && !lines.get(row).isBlank(); row++) {
                assertEquals(columns, Csv.cells(lines.get(row)).size(),
                        named + ": a row has a cell a column, under the name of it");
            }
        }
    }

    @Test
    void aRowIsARowAndAnEmptyCellIsARegisterItDoesNotSet() throws IOException {
        List<String> rows = Files.readString(Path.of("doc/tunes/circus.csv")).lines()
                .dropWhile(one -> !one.startsWith("### rows"))
                .skip(2).takeWhile(one -> !one.isBlank()).toList();
        assertEquals(4, rows.size(), "one line a row that sets something");
        List<String> first = Csv.cells(rows.get(0));
        assertEquals("0", first.get(0), "the row");
        assertEquals("163", first.get(1), "R0");
        List<String> second = Csv.cells(rows.get(1));
        assertEquals("", second.get(3), "row 1 does not set R2, so its cell is empty");
        assertEquals("12", second.get(2), "and it does set R1");
    }

    @Test
    void eachSourceOpensATableOfItsOwn() {
        Tune tune = new Tune("", "", "", 50, Tunes.repeating(List.of(
                new Row(Map.of(), Map.of(Timer.A, Tunes.struck(Tunes.setting(Register.R8),
                        Tunes.repeating("first", List.of(15, 0), 0), Prescaler.BY_4, 100))),
                new Row(Map.of(), Map.of(Timer.D, Tunes.struck(Tunes.setting(Register.R9),
                        Tunes.once("second", List.of(1, 2, 3)), Prescaler.BY_4, 100)))), 0));
        List<String> named = Csv.write(Tunes.multi(tune)).lines()
                .filter(one -> one.startsWith("###"))
                .map(one -> one.substring(3).strip()).toList();
        assertEquals(List.of("multi", "tune", "source", "value", "source", "value",
                "rows", "timerA", "timerD"), named,
                "each source opens a table, and so does each timer a row uses");
        assertEquals(Tunes.multi(tune), Csv.read(Csv.write(Tunes.multi(tune))));
    }

    @Test
    void aTuneOpensItsOwnTablesAndTheNextTuneOpensTheNext() {
        Tune one = new Tune("one", "", "", 50,
                Tunes.repeating(List.of(Tunes.row(Map.of(Register.R0, 1))), 0));
        Tune two = new Tune("two", "", "", 60,
                Tunes.repeating(List.of(Tunes.row(Map.of(Register.R0, 2))), 0));
        Multi multi = new Multi(List.of(one, two));
        String csv = Csv.write(multi);
        assertEquals(2, csv.lines().filter(said -> said.equals("### tune")).count(),
                "one table a tune opens it");
        assertTrue(!csv.contains(",tune,"), "and a table needs no column for its tune");
        assertEquals(multi, Csv.read(csv));
    }

    @Test
    void aMultiOfSeveralTunesCrossesOverTuneByTune() throws IOException {
        Multi multi = Text.read(Files.readString(Path.of("doc/tunes/two-tunes.json")));
        assertEquals(2, multi.tunes().size());
        Multi back = Csv.read(Csv.write(multi));
        assertEquals(multi, back, "the multi is the multi it was");
        for (int at = 0; at < multi.tunes().size(); at++) {
            assertEquals(multi.tunes().get(at), back.tunes().get(at),
                    "tune " + (at + 1) + " is the tune it was");
        }
        assertTrue(!multi.tunes().get(0).equals(multi.tunes().get(1)),
                "and the two are not one tune written twice");
    }

    @Test
    void twoTunesMayNameOneSourceAndTheyStayApart() throws IOException {
        Multi multi = Text.read(Files.readString(Path.of("doc/tunes/two-tunes.json")));
        Source one = Tunes.sources(multi.tunes().get(0)).get(0);
        Source two = Tunes.sources(multi.tunes().get(1)).get(0);
        assertEquals(Tunes.name(one), Tunes.name(two), "both tunes call a source this");
        assertTrue(!one.equals(two), "and the two are not one source");
        assertEquals(225, Tunes.values(one).size());
        assertEquals(293, Tunes.values(two).size());

        Multi back = Csv.read(Csv.write(multi));
        assertEquals(one, Tunes.sources(back.tunes().get(0)).get(0),
                "a source belongs to the tune whose tables it stands under, not to its name");
        assertEquals(two, Tunes.sources(back.tunes().get(1)).get(0));
    }

    @Test
    void theTablesOfOneTuneStandTogether() throws IOException {
        Multi multi = Text.read(Files.readString(Path.of("doc/tunes/two-tunes.json")));
        List<String> named = Csv.write(multi).lines().filter(one -> one.startsWith("###"))
                .map(one -> one.substring(3).strip()).toList();
        assertEquals(List.of("multi",
                "tune", "source", "value", "source", "value", "source", "value",
                        "rows", "timerD",
                "tune", "source", "value", "rows", "timerD"), named,
                "a tune opens its tables and the next tune opens the next");
    }

    @Test
    void aTimerHoldsWhatTheTextFormHoldsOfIt() {
        Source square = Tunes.repeating("square", List.of(15, 0), 0);
        Tune tune = new Tune("", "", "", 50, Tunes.repeating(List.of(
                new Row(Map.of(), Map.of(Timer.A, Tunes.struck(
                        Tunes.setting(Register.R8), square, Prescaler.BY_4, 100))),
                new Row(Map.of(), Map.of(Timer.A, Tunes.bend(Prescaler.BY_4, 90))),
                new Row(Map.of(), Map.of(Timer.A, Tunes.STOP))), 0));
        List<String> rows = Csv.write(Tunes.multi(tune)).lines()
                .dropWhile(one -> !one.startsWith("### timerA"))
                .skip(2).takeWhile(one -> !one.isBlank()).toList();
        assertEquals(List.of(
                "0,0,8,1,4,100,1,1",
                "1,1,,,4,90,0,0",
                "2,2,,,,,,"),
                rows,
                "a shape, a target and the two resets are the numbers JSON writes,"
                        + " and a cell is empty where that form writes -1");
        assertEquals(Tunes.multi(tune), Csv.read(Csv.write(Tunes.multi(tune))));
    }

    @Test
    void aTimerNoRowStatesOpensNoTable() {
        Tune tune = new Tune("", "", "", 50,
                Tunes.repeating(List.of(Tunes.row(Map.of(Register.R0, 1))), 0));
        assertTrue(!Csv.write(Tunes.multi(tune)).contains("### timer"),
                Csv.write(Tunes.multi(tune)));
    }

    @Test
    void everyShapeCrossesOver() {
        Source square = Tunes.repeating("square 15", List.of(15, 0), 0);
        Source drum = Tunes.once("drum", List.of(8, 12, 15, 13, 5));
        Map<Register, Integer> all = new java.util.EnumMap<>(Register.class);
        for (Register register : Register.values()) {
            all.put(register, Chip.most(register));
        }
        Tune tune = new Tune("a tune", "a composer", "CsvTest", 50, Tunes.repeating(List.of(
                Tunes.row(all),
                new Row(Map.of(), Map.of(
                        Timer.A, Tunes.struck(Tunes.setting(Register.R8), square,
                                Prescaler.BY_4, 122),
                        Timer.D, new Start(Tunes.setting(Register.R10), drum,
                                Prescaler.BY_200, 256, false, true))),
                new Row(Map.of(), Map.of(Timer.A, Tunes.bend(Prescaler.BY_4, 118))),
                new Row(Map.of(Register.R8, 12), Map.of(Timer.A, Tunes.STOP)),
                Tunes.EMPTY), 1));
        Multi multi = Tunes.multi(tune);
        assertEquals(multi, Csv.read(Csv.write(multi)));
        assertEquals(multi, Text.read(Text.write(multi)), "and JSON reads it back too");
    }

    @Test
    void aRowThatSetsNoRegisterIsLeftOutOfTheTable() {
        Tune tune = new Tune("", "", "", 50, Tunes.repeating(List.of(
                Tunes.row(Map.of(Register.R7, 56)), Tunes.EMPTY,
                Tunes.row(Map.of(Register.R7, 49))), 0));
        String csv = Csv.write(Tunes.multi(tune));
        List<String> rows = csv.lines().dropWhile(one -> !one.startsWith("### rows"))
                .skip(2).takeWhile(one -> !one.isBlank()).toList();
        assertEquals(List.of("0,,,,,,,,56,,,,,,", "2,,,,,,,,49,,,,,,"), rows,
                "the row column is the row number, so a row that sets none is left out");
        assertEquals(Tunes.multi(tune), Csv.read(csv), "and it reads back to three rows");
    }

    @Test
    void aCellWithACommaOrAQuoteInItIsQuoted() {
        Tune tune = new Tune("a, \"quoted\", title", "", "", 50,
                Tunes.repeating(List.of(Tunes.EMPTY), 0));
        String csv = Csv.write(Tunes.multi(tune));
        assertTrue(csv.contains("\"a, \"\"quoted\"\", title\""), csv);
        assertEquals(tune.title(), Csv.read(csv).tunes().get(0).title());
    }

    @Test
    void aCellWithALineFeedInItIsAnError() {
        Tune tune = new Tune("a\ntitle", "", "", 50,
                Tunes.repeating(List.of(Tunes.EMPTY), 0));
        IllegalArgumentException no = assertThrows(IllegalArgumentException.class,
                () -> Csv.write(Tunes.multi(tune)));
        assertTrue(String.valueOf(no.getMessage()).contains("a line feed"),
                String.valueOf(no.getMessage()));
    }

    @Test
    void aTextOfAnotherFormatOrVersionIsTurnedAway() {
        assertThrows(IllegalArgumentException.class,
                () -> Csv.read("### multi\nformat,version,tunes\nymxr,1,0\n"));
        assertThrows(IllegalArgumentException.class,
                () -> Csv.read("### multi\nformat,version,tunes\nymxs,9,0\n"));
    }

    /** A file of the form this replaced, where the name and the columns
     *  stood on one line, and a table whose name is its last line. */
    @Test
    void aHeadingThisDoesNotReadIsNamed() {
        IllegalArgumentException old = assertThrows(IllegalArgumentException.class,
                () -> Csv.read("### multi,format,version,tunes\nymxs,1,1\n"));
        assertTrue(String.valueOf(old.getMessage()).contains("has a comma in it"),
                String.valueOf(old.getMessage()));
        IllegalArgumentException alone = assertThrows(IllegalArgumentException.class,
                () -> Csv.read("### multi\n"));
        assertTrue(String.valueOf(alone.getMessage()).contains("names no columns"),
                String.valueOf(alone.getMessage()));
        IllegalArgumentException last = assertThrows(IllegalArgumentException.class,
                () -> Csv.read("### multi\nformat,version,tunes\nymxs,1,1\n\n### tune\n"));
        assertTrue(String.valueOf(last.getMessage()).contains("names no columns"),
                String.valueOf(last.getMessage()));
    }

    @Test
    void aTextWithNoMultiTableIsTurnedAway() {
        IllegalArgumentException no = assertThrows(IllegalArgumentException.class,
                () -> Csv.read("### tune\ntune\n1\n"));
        assertTrue(String.valueOf(no.getMessage()).contains("multi"),
                String.valueOf(no.getMessage()));
    }

    @Test
    void aColumnIsFoundByItsNameAndNotItsPlace() {
        String csv = """
                ### multi
                version,tunes,format
                1,1,ymxs

                ### tune
                frames,rate,repeat,writer,composer,title
                1,50,0,a writer,a composer,a title

                ### rows
                r0,row
                200,0
                """;
        Tune tune = Csv.read(csv).tunes().get(0);
        assertEquals("a title", tune.title());
        assertEquals(50, tune.rate());
        assertEquals(200, Tunes.rows(tune).get(0).registers().get(Register.R0));
    }
}

File: go/csv/csv_test.go
package csv_test

import (
	"os"
	"strings"
	"testing"

	"github.com/odipar/ymxs/go/csv"
	"github.com/odipar/ymxs/go/text"
)

var tunes = []string{"circus", "digidrum", "example", "retrigger", "turrican-2", "two-tunes"}

// The two forms write the same tune, so a tune that crosses to the tables
// and back is the JSON it was.
func TestATuneCrossesToTheTablesAndBack(t *testing.T) {
	for _, named := range tunes {
		said, err := os.ReadFile("../../doc/tunes/" + named + ".json")
		if err != nil {
			t.Fatal(err)
		}
		multi, err := text.Read(string(said))
		if err != nil {
			t.Fatalf("%s: %v", named, err)
		}
		back, err := csv.Read(csv.Write(multi))
		if err != nil {
			t.Fatalf("%s: %v", named, err)
		}
		if again := text.Write(back); again != string(said) {
			t.Errorf("%s: the tune moved crossing to the tables and back", named)
		}
	}
}

// The tables of doc/tunes/circus.csv are the tables this writes.
func TestTheDocumentsTablesAreTheTablesThisWrites(t *testing.T) {
	said, err := os.ReadFile("../../doc/tunes/circus.json")
	if err != nil {
		t.Fatal(err)
	}
	tables, err := os.ReadFile("../../doc/tunes/circus.csv")
	if err != nil {
		t.Fatal(err)
	}
	multi, err := text.Read(string(said))
	if err != nil {
		t.Fatal(err)
	}
	if written := csv.Write(multi); written != string(tables) {
		t.Errorf("the tables moved:\n%s", written)
	}
}

func TestALineOfCellsReadsBackTheCellsItHasInIt(t *testing.T) {
	for _, one := range []struct {
		line  string
		cells []string
	}{
		{"a,b,c", []string{"a", "b", "c"}},
		{`a,"b,c",d`, []string{"a", "b,c", "d"}},
		{`"he said ""no"""`, []string{`he said "no"`}},
		{"a,,c", []string{"a", "", "c"}},
	} {
		cells := csv.Cells(one.line)
		if len(cells) != len(one.cells) {
			t.Errorf("%s reads as %d cells, and %d are in it", one.line, len(cells),
				len(one.cells))
			continue
		}
		for at, said := range one.cells {
			if cells[at] != said {
				t.Errorf("%s: cell %d is %q, and %q is in it", one.line, at,
					cells[at], said)
			}
		}
	}
}

// A heading line names the table alone, and the line after it names the
// columns, so a column name is over the cells it names.
func TestAHeadingNamesTheTableAndTheLineAfterItTheColumns(t *testing.T) {
	tables, err := os.ReadFile("../../doc/tunes/circus.csv")
	if err != nil {
		t.Fatal(err)
	}
	lines := strings.Split(strings.TrimRight(string(tables), "\n"), "\n")
	opened := 0
	for at, line := range lines {
		if !strings.HasPrefix(line, "###") {
			continue
		}
		opened++
		named := strings.TrimSpace(line[3:])
		if strings.Contains(named, ",") {
			t.Errorf("the heading %q names more than the table", line)
			continue
		}
		columns := len(csv.Cells(lines[at+1]))
		for row := at + 2; row < len(lines) && strings.TrimSpace(lines[row]) != ""; row++ {
			if cells := len(csv.Cells(lines[row])); cells != columns {
				t.Errorf("%s: row %d has %d cells under %d columns", named, row,
					cells, columns)
			}
		}
	}
	if opened != 3 {
		t.Errorf("circus is rows alone, so it opens 3 tables and %d were read", opened)
	}
}

// A file of the form this replaced, and a table whose name is its last
// line, are both named rather than read as something else.
func TestAHeadingThisDoesNotReadIsNamed(t *testing.T) {
	for _, one := range []struct {
		said string
		of   string
	}{
		{"### multi,format,version,tunes\nymxs,1,1\n", "has a comma in it"},
		{"### multi\n", "names no columns"},
		{"### multi\nformat,version,tunes\nymxs,1,1\n\n### tune\n", "names no columns"},
	} {
		_, err := csv.Read(one.said)
		if err == nil {
			t.Errorf("%q reads, and it is not this form", one.said)
			continue
		}
		if !strings.Contains(err.Error(), one.of) {
			t.Errorf("%q reports %q, and %q is what is wrong with it", one.said,
				err, one.of)
		}
	}
}

// TestASourceOfSeveralValuesCrossesTheForms is a tune with a source of
// three values a row: it crosses to the tables and back, and the value
// block names value1, value2 and value3 (csv.md 3.6).
func TestASourceOfSeveralValuesCrossesTheForms(t *testing.T) {
	said := `{"format":"ymxs","version":4,"tunes":[{"title":"","composer":"",
		"writer":"t","rate":50,"rows":2,"repeat":0,
		"sources":[{"name":"a sweep","repeat":0,"values":[[46,1,15],[32,1,13]]}],
		"registers":{"r7":[56,-1]},
		"timerA":{"shape":[0,-1],"target":[17,-1],"source":[1,-1],
		"prescaler":[50,-1],"count":[60,-1],"timerReset":[1,-1],"placeReset":[1,-1]}}]}`
	multi, err := text.Read(said)
	if err != nil {
		t.Fatal(err)
	}
	made := csv.Write(multi)
	if !strings.Contains(made, "row,value1,value2,value3") {
		t.Errorf("the block names the cells it has:\n%s", made)
	}
	if !strings.Contains(made, "0,46,1,15") {
		t.Errorf("and the row is its values:\n%s", made)
	}
	back, err := csv.Read(made)
	if err != nil {
		t.Fatal(err)
	}
	if text.Write(back) != text.Write(multi) {
		t.Errorf("the tune moved crossing to the tables and back:\n%s\n%s",
			text.Write(multi), text.Write(back))
	}
}

// A file of version 3 has one value a row in every source and a target of
// 0 to 13 (json.md 2.4): such a file reads, and a row of several values or
// a target above 13 in one is an error of the form, one line of csv.md 5.1.
func TestAFileOfTheVersionBeforeThisOneReadsItsShapes(t *testing.T) {
	three := `### multi
format,version,tunes
ymxs,3,1

### tune
title,composer,writer,rate,rows,repeat
,,t,50,2,0

### source
name,repeat
a tone,0

### value
row,value
0,13
1,0

### timerA
row,shape,source,target,prescaler,count,timerReset,placeReset
0,0,1,8,50,60,1,1
`
	if _, err := csv.Read(three); err != nil {
		t.Fatalf("a file of version 3 reads: %v", err)
	}
	for _, one := range []struct{ from, to, line string }{
		{"row,value\n0,13\n1,0", "row,value1,value2\n0,13,1\n1,0,1",
			"source a tone has a row of several values, and version 3 has one value a row"},
		{"0,0,1,8,", "0,0,1,14,", "target 14, and version 3 reaches 0 to 13"},
	} {
		_, err := csv.Read(strings.Replace(three, one.from, one.to, 1))
		if err == nil || err.Error() != one.line {
			t.Errorf("%q reports %v, and csv.md 5.1 has %q", one.to, err, one.line)
		}
	}
}
